hudi-common/src/main/java/org/apache/hudi/io/storage/HoodieParquetKeyedLookupReader.java

```diff
@@ -0,0 +1,318 @@
+      }
+      LOG.info("Done with rowGroup");
+    }
+    return keyToValue;
+  }
+
+  public static boolean shouldSkip(RowGroup rowGroup, int colNumber, Queue<String> keys) {
+    if (rowGroup.getBloomFilters().isEmpty()) {
+      return false;
+    }
+
+    BloomFilter bloomFilter = null;
+    try {
+      bloomFilter = rowGroup.getBloomFilters().get(colNumber);
+    } catch (Exception e) {
+      LOG.warn("Can not load the bloom filter correctly.", e);
+      return false;
+    }
+
+    // Bloom filter is not found for this column; can not skip.
+    if (bloomFilter == null) {
+      return false;
+    }
+
+    while (!keys.isEmpty()) {
+      String key = keys.poll();
+      Binary binary = Binary.fromString(key);
+      long hash = bloomFilter.hash(binary);
+      // At least one key is found; can not skip.
+      if (bloomFilter.findHash(hash)) {
+        return false;
+      }
+    }
+
+    // No keys are found in the filter; skip.
+    return true;
+  }
+
+  private int searchColumn(RowGroup rowGroup, String columnDotPath) {
+    final int numColumns = rowGroup.getBlockMetaData().getColumns().size();
+    for (int i = 0; i < numColumns; i++) {
+      if (rowGroup.getBlockMetaData().getColumns().get(i).getPath().toDotString().equals(columnDotPath)) {
+        return i;
+      }
+    }
+    return -1;
+  }
+
+  private ArrayList<Pair<String, Long>> lookupKeyColumnChunk(CompressionConverter.TransParquetFileReader reader,
+                                                             RowGroup rowGroup,
+                                                             int colNumber,
+                                                             Queue<String> keys) throws IOException {
+    ColumnChunkMetaData chunkMetaData = rowGroup.getBlockMetaData().getColumns().get(colNumber);
+    ColumnIndex columnIndex = rowGroup.getColumnIndices().get(colNumber);
+    OffsetIndex pageLocation = rowGroup.getPageLocations().get(colNumber);
+    CompressionCodecFactory.BytesInputDecompressor decompressor = codecFactory.getDecompressor(chunkMetaData.getCodec());
+    ColumnPath columnPath = rowGroup.getBlockMetaData().getColumns().get(colNumber).getPath();
+    ColumnDescriptor columnDescriptor = metadataLoader.getFileMetaData().getSchema().getColumnDescription(columnPath.toArray());
+    PrimitiveType type = metadataLoader.getFileMetaData().getSchema().getType(columnDescriptor.getPath()).asPrimitiveType();
+
+    AtomicInteger pageCounter = new AtomicInteger(0);
+    int totalPages = pageLocation.getPageCount();
+    ArrayList<Pair<String, Long>> keyAndPositions = new ArrayList<>();
+    for (int pageIndex = 0; pageIndex < totalPages; pageIndex++) {
+      String pageMinKey = new String(columnIndex.getMinValues().get(pageIndex).array(), StandardCharsets.UTF_8);
+      String pageMaxKey = new String(columnIndex.getMaxValues().get(pageIndex).array(), StandardCharsets.UTF_8);
+
+      while (keys.peek() != null && keys.peek().compareTo(pageMinKey) < 0) {
```

Comment:
can we bring back some of the comments I originally had. It will help readers down the line, maintain the code.

Reply:
Done!